From 161b9b8a94955decf1acfae1837d0bce0dcc29e3 Mon Sep 17 00:00:00 2001
From: Jack Ding <jack.ding@windriver.com>
Date: Mon, 24 Apr 2017 16:08:24 -0400
Subject: [PATCH 21/53] Pike Rebase: Fix boot from snapshot with
 block_device_mapping

Inst fail to boot with vol from snapshot with block_device_mapping

Instance failed to boot from volume which was created from snapshot
of an instance that has block_device_mapping in metadata.

Error in nova-api.log:
HTTP exception thrown: Invalid image metadata.
Error: A list is required in field img_block_device_mapping, not a unicode

---
 cinder/volume/api.py | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/cinder/volume/api.py b/cinder/volume/api.py
index 8e4625e..4575228 100644
--- a/cinder/volume/api.py
+++ b/cinder/volume/api.py
@@ -31,6 +31,7 @@ import os
 
 from oslo_config import cfg
 from oslo_log import log as logging
+from oslo_serialization import jsonutils
 from oslo_utils import excutils
 from oslo_utils import strutils
 from oslo_utils import timeutils
@@ -1285,6 +1286,8 @@ class API(base.Base):
         db_data = self.db.volume_glance_metadata_get_all(context)
         results = collections.defaultdict(dict)
         for meta_entry in db_data:
+            if meta_entry['key'] == 'block_device_mapping':
+                meta_entry['value'] = jsonutils.loads(meta_entry['value'])
             results[meta_entry['volume_id']].update({meta_entry['key']:
                                                      meta_entry['value']})
         return results
@@ -1301,6 +1304,8 @@ class API(base.Base):
                                                           volume_id_list)
         results = collections.defaultdict(dict)
         for meta_entry in db_data:
+            if meta_entry['key'] == 'block_device_mapping':
+                meta_entry['value'] = jsonutils.loads(meta_entry['value'])
             results[meta_entry['volume_id']].update({meta_entry['key']:
                                                      meta_entry['value']})
         return results
-- 
2.7.4

